File: ios/Classes/ZebraRfidReaderSdkPlugin.swift
import Flutter
import UIKit

public class ZebraRfidReaderSdkPlugin: NSObject, FlutterPlugin {
  var zebraConnectionHelper: ZebraConnectionHelper!
  var apiDelegate: SdkApiDelegate!
  var tagHandlerEvent: TagDataEventHandler!
  var readTagEvent: TagDataEventHandler!
  var tagFindingEvent: TagDataEventHandler!
  let api: srfidISdkApi = ZebraApi.shared

  public static func register(with registrar: FlutterPluginRegistrar) {
    let channel = FlutterMethodChannel(name: "borda/zebra_rfid_reader_sdk", binaryMessenger: registrar.messenger())
    let instance = ZebraRfidReaderSdkPlugin()
    registrar.addMethodCallDelegate(instance, channel: channel)
    instance.initializeSdkComponents()
    instance.setupEventChannels(with: registrar.messenger())
   
  }

  private func setupEventChannels(with messenger: FlutterBinaryMessenger) {
    let tagHandlerEventChannel = FlutterEventChannel(name: "tagHandlerEvent", binaryMessenger: messenger)
    tagHandlerEventChannel.setStreamHandler(tagHandlerEvent)
    
    let readTagEventChannel = FlutterEventChannel(name: "readTagEvent", binaryMessenger: messenger)
    readTagEventChannel.setStreamHandler(readTagEvent)

    let tagFindingEventChannel = FlutterEventChannel(name: "tagFindingEvent", binaryMessenger: messenger)
    tagFindingEventChannel.setStreamHandler(tagFindingEvent)
  }

  private func initializeSdkComponents() {
    tagHandlerEvent = TagDataEventHandler()
    readTagEvent = TagDataEventHandler()
    tagFindingEvent = TagDataEventHandler()
    
    apiDelegate = SdkApiDelegate(api: api, tagHandlerEvent: tagHandlerEvent, readTagEvent: readTagEvent, tagFindingEvent: tagFindingEvent)
    
    zebraConnectionHelper = ZebraConnectionHelper(api: api, delegate: apiDelegate, tagHandlerEvent: tagHandlerEvent, tagFindingEvent: tagFindingEvent, readTagEvent: readTagEvent)
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    switch call.method {
 case "getAvailableReaderList":
                self.zebraConnectionHelper.getAvailableReaderList(api: self.api, result: result)
            case  "disconnect" :
                self.zebraConnectionHelper.disconnect()

            case "setAntennaPower":
                    if let transmitPowerIndex = call.arguments as? [String: Any],
                       let powerIndex = transmitPowerIndex["transmitPowerIndex"] as? Int {
                        self.zebraConnectionHelper.setAntennaPower(transmitPowerIndex: powerIndex)
                    } else {
                        result(FlutterError(code: "INVALID_ARGUMENT", message: "Missing or invalid argument: transmitPowerIndex", details: nil))
                    }

            case "setDynamicPower":
                    if let args = call.arguments as? [String: Any],
                       let isEnable = args["isEnable"] as? Bool {
                        self.zebraConnectionHelper.setDynamicPower(isEnable: isEnable)
                    } else {
                        result(FlutterError(code: "INVALID_ARGUMENT", message: "Missing or invalid argument: isEnable", details: nil))
                    }
                case "setBeeperVolume":
                    if let args = call.arguments as? [String: Any],
                       let level = args["level"] as? Int {
                        self.zebraConnectionHelper.setBeeperVolumeConfig(level:level)
                    } else {
                        result(FlutterError(code: "INVALID_ARGUMENT", message: "Missing or invalid argument: level", details: nil))
                    }
            case "connect":
                if let args = call.arguments as? [String: Any], let name = args["name"] as? String, let readerConfig = args["readerConfig"] as? [String: Any] {
                    print("Flutter -> connect called with name: \(name)")
                    print("Flutter -> USER CONFIG: \(readerConfig)")
                    self.zebraConnectionHelper.connect(name: name, readerConfig: readerConfig)
                } else {
                    result(FlutterError(code: "BAD_ARGS", message: "Required arguments 'name' or 'readerConfig' not provided", details: nil))
                }
            case "findTheTag":
                if let args = call.arguments as? [String: Any], let tag = args["tag"] as? String {
                    self.zebraConnectionHelper.findTheTag(tag: tag)
                } else {
                    result(FlutterError(code: "BAD_ARGS", message: "No tag provided", details: nil))
                }
            case "stopFindingTheTag":
                self.zebraConnectionHelper.stopFindingTheTag()
            default:
                result(FlutterMethodNotImplemented)
            }
  }
}

File: android/src/main/kotlin/com/borda/zebra_rfid_reader_sdk/ZebraRfidReaderSdkPlugin.kt
package com.borda.zebra_rfid_reader_sdk

import android.util.Log
import com.borda.zebra_rfid_reader_sdk.utils.BordaReaderDevice
import com.borda.zebra_rfid_reader_sdk.utils.ConnectionStatus
import com.borda.zebra_rfid_reader_sdk.utils.LOG_TAG
import com.google.gson.Gson
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result

/** ZebraRfidReaderSdkPlugin */
class ZebraRfidReaderSdkPlugin : FlutterPlugin, MethodCallHandler {
    /// The MethodChannel that will the communication between Flutter and native Android
    ///
    /// This local reference serves to register the plugin with the Flutter Engine and unregister it
    /// when the Flutter Engine is detached from the Activity
    private lateinit var methodChannel: MethodChannel
    private lateinit var connectionHelper: ZebraConnectionHelper

    private lateinit var eventChannel: EventChannel
    private lateinit var tagFindingEvent: EventChannel
    private lateinit var tagDataEventHandler: TagDataEventHandler


    override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        methodChannel =
            MethodChannel(flutterPluginBinding.binaryMessenger, "borda/zebra_rfid_reader_sdk")
        methodChannel.setMethodCallHandler(this)

        eventChannel = EventChannel(flutterPluginBinding.binaryMessenger, "tagHandlerEvent")
        tagFindingEvent = EventChannel(flutterPluginBinding.binaryMessenger, "tagFindingEvent")

        tagDataEventHandler = TagDataEventHandler()

        eventChannel.setStreamHandler(tagDataEventHandler)
        tagFindingEvent.setStreamHandler(tagDataEventHandler)
        Log.d(LOG_TAG, "onAttachedToEngine called")
        connectionHelper =
            ZebraConnectionHelper(flutterPluginBinding.applicationContext, this::emit)
    }

    override fun onMethodCall(call: MethodCall, result: Result) =
        when (call.method) {
            "connect" -> {
                Log.d(LOG_TAG, "connect called")
                val name = call.argument<String>("name")!!
                val readerConfig = call.argument<HashMap<String, Any>>("readerConfig")!!
                Log.d(LOG_TAG, "will try to connect to -> $name")
                Log.d(LOG_TAG, "USER CONFIG -> $readerConfig")
                connectionHelper.connect(name, readerConfig)
            }

            "disconnect" -> {
                Log.d(LOG_TAG, "disconnect called")
                connectionHelper.disconnect()
            }

            "setAntennaPower" -> {
                val transmitPowerIndex = call.argument<Int>("transmitPowerIndex")!!
                connectionHelper.setAntennaConfig(transmitPowerIndex)
            }

            "setDynamicPower" -> {
                val isEnable = call.argument<Boolean>("isEnable")!!
                connectionHelper.setDynamicPower(isEnable)
            }

            "setBeeperVolume" -> {
                val level = call.argument<Int>("level")!!
                connectionHelper.setBeeperVolumeConfig(level)
            }

            "findTheTag" -> {
                val tag = call.argument<String>("tag")!!
                Log.d("ENGIN", "findTheTag called with tag -> $tag")
                connectionHelper.findTheTag(tag)
            }

            "stopFindingTheTag" -> {
                connectionHelper.stopFindingTheTag()
            }

            "getAvailableReaderList" -> {
                Log.d(LOG_TAG, "getAvailableReaderList called")
                val readers = connectionHelper.getAvailableReaderList()
                val dataList = mutableListOf<BordaReaderDevice>()
                for (reader in readers) {
                    val device = BordaReaderDevice(
                        ConnectionStatus.notConnected,
                        reader.name.toString(),
                        null,
                    )
                    dataList.add(device)
                }

                result.success(Gson().toJson(dataList))
            }

            else -> result.notImplemented()
        }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        methodChannel.setMethodCallHandler(null)
        eventChannel.setStreamHandler(null)
        connectionHelper.dispose()
    }


    private fun emit(tagData: String) {
        tagDataEventHandler.sendEvent(tagData)
    }
}
